index.js

```diff
@@ -323,139 +327,251 @@ Hyperlog.prototype.batch = function (docs, opts, cb) {
   var id = opts.log || self.id
   opts.log = id
 
-  var nodes = new Array(docs.length)
-  var logLinks = new Array(docs.length)
-  var batch = [] // dynamic length
+  var logLinks = {}
+  var lockRelease = null
+  var latestSeq
 
-  var onlocked = function (release) {
-    self.ready(function () {
-      self.logs.head(id, function (err, seq) {
-        if (err) return release(cb, err)
-        else batchAndRelease(release, seq)
-      })
-    })
+  // Bubble up errors on non-batch (1 element) calls.
+  var bubbleUpErrors = false
+  if (docs.length === 1) {
+    bubbleUpErrors = true
   }
 
-  var batchAndRelease = function (release, seq) {
-    // Once all batch insertion operations are computed, perform a leveldb
-    // batch insert, and update self.changes to reflect the new largest change#
-    // in the hyperlog.
-    var next = after(function (err) {
-      if (err) return release(cb, err)
-
-      self.db.batch(batch, function (err) {
-        if (err) return release(cb, err)
-
-        self.changes = nodes.reduce(maxChange, self.changes)
+  // 1. construct initial hyperlog "node" for each of "docs"
+  var nodes = docs.map(function (doc) {
+    return constructInitialNode(doc, opts)
+  })
 
-        for (var i = 0; i < nodes.length; i++) self.emit('add', nodes[i])
+  // 2. emit all preadd events
+  nodes.forEach(function (node) {
+    self.emit('preadd', node)
+  })
 
-        release(cb, null, nodes)
+  waterfall([
+    // 3. lock the hyperlog (if needed)
+    // 4. wait until the hyperlog is 'ready'
+    // 5. retrieve the seq# of this hyperlog's head
+    lockAndGetSeqNumber,
+
+    // 3. hash (async/sync) all nodes
+    // 4. retrieve + set 'getLinks' for each node
+    function (seq, release, done) {
+      lockRelease = release
+      latestSeq = seq
+
+      hashNodesAndFindLinks(nodes, done)
+    },
+
+    // 8. dedupe the node against the params AND the hyperlog (in sequence)
+    function (nodes, done) {
+      dedupeNodes(nodes, latestSeq, done)
+    },
+
+    // 9. create each node's leveldb batch operation object
+    function (nodes, done) {
+      computeBatchNodeOperations(nodes, done)
+    },
+
+    // 10. perform the leveldb batch op
+    function (nodes, batchOps, done) {
+      self.db.batch(batchOps, function (err) {
+        if (err) {
+          nodes.forEach(rejectNode)
+          return done(err)
+        }
+        done(null, nodes)
       })
-    })
+    },
 
-    var added = nodes.length > 1 ? {} : null
-    var seqIdx = 1
-    var changeIdx = 1
-    var batchLock = mutexify()
+    // 11. update the hyperlog's change#
+    // 12. emit all add/reject events
+    function (nodes, done) {
+      self.changes = nodes.reduce(maxChange, self.changes)
+      done(null, nodes)
+    }
+  ], function (err, nodes) {
+    // release lock, if necessary
+    if (lockRelease) {
+      return lockRelease(onUnlocked, err)
+    } else {
```

Comment:
Sure. I usually like to include `else` statements in this instances to make it explicit that both cases are being handled in different ways. Sometimes future devs remove `return` statements without noticing that it will cause fallthrough with unintended side effects.